Sources/BSON/BSONDocumentIterator.swift

```diff
@@ -28,13 +28,25 @@ public struct BSONDocumentIterator: IteratorProtocol {
 
     /// Advances to the next element and returns it, or nil if no next element exists.
     public mutating func next() -> BSONDocument.KeyValuePair? {
+        do {
+            return try self._next()
+        } catch {
+            fatalError("\(error)")
+        }
+    }
+
+    /**
+     * Advances to the next element and returns it, or nil if no next element exists.
+     * - Throws:
+     */
+    internal mutating func _next() throws -> BSONDocument.KeyValuePair? {
         guard self.buffer.readableBytes != 0 else {
             // Iteration has been exhausted
             return nil
         }
 
         guard let typeByte = self.buffer.readInteger(as: UInt8.self) else {
-            fatalError("BSONDocumentIterator Failed: Cannot read from \(self.buffer)")
+            throw BSONError.InternalError(message: "BSONDocumentIterator Failed: Cannot read from \(self.buffer)")
```

Comment:
I think all of the error message in this method might be more helpful if we could include the byte offset at which we found invalid BSON